// node-coap-client/build/lib/SocketWrapper.js
import { EventEmitter } from "events";
import { dtls } from "node-dtls-client";
import { getSocketAddressFromURLSafeHostnameWithoutLookup } from "./Hostname.js";
export class SocketWrapper extends EventEmitter {
    socket;
    isDtls;
    isClosed;
    constructor(socket) {
        super();
        this.socket = socket;
        this.isDtls = (socket instanceof dtls.Socket);
        socket
            .on("message", (message, rinfo) => {
            this.emit("message", message, rinfo);
        })
            .on("error", (err) => {
            this.emit("error", err);
        })
            .on("close", () => {
            this.emit("close");
        });
    }
    send(msg, origin) {
        if (this.isClosed)
            return;
        if (this.isDtls) {
            this.socket.send(msg);
        }
        else {
            this.socket.send(msg, origin.port, getSocketAddressFromURLSafeHostnameWithoutLookup(origin.hostname));
        }
    }
    close() {
        if (this.isClosed)
            return;
        this.isClosed = true;
        if (this.isDtls) {
            this.socket.close();
        }
        else {
            this.socket.close();
        }
    }
}
