// node-dtls-client/build/DTLS/DTLSPlaintext.js
import { ContentType } from "../TLS/ContentType.js";
import { ProtocolVersion } from "../TLS/ProtocolVersion.js";
import { TLSStruct } from "../TLS/TLSStruct.js";
import * as TypeSpecs from "../TLS/TypeSpecs.js";
export class DTLSPlaintext extends TLSStruct {
    type;
    version;
    epoch;
    sequence_number;
    fragment;
    static __spec = {
        type: TypeSpecs.define.Struct(ContentType),
        version: TypeSpecs.define.Struct(ProtocolVersion),
        epoch: TypeSpecs.uint16,
        sequence_number: TypeSpecs.uint48,
        // length field is implied in the variable length vector
        fragment: TypeSpecs.define.Buffer(0, 2 ** 14),
    };
    static spec = TypeSpecs.define.Struct(DTLSPlaintext);
    constructor(type, version = new ProtocolVersion(), epoch, sequence_number, fragment) {
        super(DTLSPlaintext.__spec);
        this.type = type;
        this.version = version;
        this.epoch = epoch;
        this.sequence_number = sequence_number;
        this.fragment = fragment;
    }
    static createEmpty() {
        return new DTLSPlaintext(null, null, null, null, null);
    }
}
