// node-dtls-client/build/DTLS/DTLSCompressed.js
import { ContentType } from "../TLS/ContentType.js";
import { ProtocolVersion } from "../TLS/ProtocolVersion.js";
import { TLSStruct } from "../TLS/TLSStruct.js";
import * as TypeSpecs from "../TLS/TypeSpecs.js";
import { DTLSPlaintext } from "./DTLSPlaintext.js";
export class DTLSCompressed extends TLSStruct {
    type;
    version;
    epoch;
    sequence_number;
    fragment;
    static __spec = {
        type: ContentType.__spec,
        version: TypeSpecs.define.Struct(ProtocolVersion),
        epoch: TypeSpecs.uint16,
        sequence_number: TypeSpecs.uint48,
        // length field is implied in the variable length vector
        fragment: TypeSpecs.define.Buffer(0, 1024 + 2 ** 14),
    };
    static spec = TypeSpecs.define.Struct(DTLSCompressed);
    constructor(type, version = new ProtocolVersion(), epoch, sequence_number, fragment) {
        super(DTLSCompressed.__spec);
        this.type = type;
        this.version = version;
        this.epoch = epoch;
        this.sequence_number = sequence_number;
        this.fragment = fragment;
    }
    static createEmpty() {
        return new DTLSCompressed(null, null, null, null, null);
    }
    /**
     * Compresses the given plaintext packet
     * @param packet - The plaintext packet to be compressed
     * @param compressor - The compressor function used to compress the given packet
     */
    static compress(packet, compressor) {
        return new DTLSCompressed(packet.type, packet.version, packet.epoch, packet.sequence_number, compressor(packet.fragment));
    }
    /**
     * Decompresses this packet into a plaintext packet
     * @param decompressor - The decompressor function used to decompress this packet
     */
    decompress(decompressor) {
        return new DTLSPlaintext(this.type, this.version, this.epoch, this.sequence_number, decompressor(this.fragment));
    }
    /**
     * Computes the MAC header representing this packet. The MAC header is the input buffer of the MAC calculation minus the actual fragment buffer.
     */
    computeMACHeader() {
        return (new MACHeader(this.epoch, this.sequence_number, this.type, this.version, this.fragment.length)).serialize();
    }
}
export class MACHeader extends TLSStruct {
    epoch;
    sequence_number;
    type;
    version;
    fragment_length;
    static __spec = {
        epoch: TypeSpecs.uint16,
        sequence_number: TypeSpecs.uint48,
        type: ContentType.__spec,
        version: TypeSpecs.define.Struct(ProtocolVersion),
        fragment_length: TypeSpecs.uint16,
    };
    constructor(epoch, sequence_number, type, version, fragment_length) {
        super(MACHeader.__spec);
        this.epoch = epoch;
        this.sequence_number = sequence_number;
        this.type = type;
        this.version = version;
        this.fragment_length = fragment_length;
    }
    static createEmpty() {
        return new MACHeader(null, null, null, null, null);
    }
}
