// node-dtls-client/build/TLS/ConnectionState.js
import { CipherSuites } from "../DTLS/CipherSuites.js";
import { ProtocolVersion } from "../TLS/ProtocolVersion.js";
import { PRF } from "./PRF.js";
import * as TypeSpecs from "./TypeSpecs.js";
export var CompressionMethod;
(function (CompressionMethod) {
    CompressionMethod[CompressionMethod["null"] = 0] = "null";
})(CompressionMethod || (CompressionMethod = {}));
(function (CompressionMethod) {
    CompressionMethod.spec = TypeSpecs.define.Enum("uint8", CompressionMethod);
})(CompressionMethod || (CompressionMethod = {}));
const master_secret_length = 48;
// const client_random_length = 32;
// const server_random_length = 32;
export class ConnectionState {
    // This doesn't seem to be used:
    // constructor(values?: Partial<ConnectionState>) {
    // 	if (values) {
    // 		for (const [key, value] of entries(values)) {
    // 			if (this.hasOwnProperty(key)) (this as any)[key] = value;
    // 		}
    // 	}
    // }
    entity = "client";
    cipherSuite = CipherSuites.TLS_NULL_WITH_NULL_NULL;
    protocolVersion = new ProtocolVersion(~1, ~0); // default to DTLSv1.0 during handshakes
    compression_algorithm = CompressionMethod.null;
    master_secret;
    client_random;
    server_random;
    key_material;
    _cipher;
    get Cipher() {
        if (this._cipher == undefined) {
            this._cipher = this.cipherSuite.specifyCipher(this.key_material, this.entity);
        }
        return this._cipher;
    }
    _decipher;
    get Decipher() {
        if (this._decipher == undefined) {
            this._decipher = this.cipherSuite.specifyDecipher(this.key_material, this.entity);
        }
        return this._decipher;
    }
    /**
     * Compute the master secret from a given premaster secret
     * @param preMasterSecret - The secret used to calculate the master secret
     * @param clientHelloRandom - The random data from the client hello message
     * @param serverHelloRandom - The random data from the server hello message
     */
    computeMasterSecret(preMasterSecret) {
        this.master_secret = PRF[this.cipherSuite.prfAlgorithm](preMasterSecret.serialize(), "master secret", Buffer.concat([this.client_random, this.server_random]), master_secret_length);
        // now we can compute the key material
        this.computeKeyMaterial();
    }
    /**
     * Calculates the key components
     */
    computeKeyMaterial() {
        const keyBlock = PRF[this.cipherSuite.prfAlgorithm](this.master_secret, "key expansion", Buffer.concat([this.server_random, this.client_random]), 2 * (this.cipherSuite.MAC.keyAndHashLength + this.cipherSuite.Cipher.keyLength + this.cipherSuite.Cipher.fixedIvLength));
        let offset = 0;
        function read(length) {
            const ret = keyBlock.slice(offset, offset + length);
            offset += length;
            return ret;
        }
        this.key_material = {
            client_write_MAC_key: read(this.cipherSuite.MAC.keyAndHashLength),
            server_write_MAC_key: read(this.cipherSuite.MAC.keyAndHashLength),
            client_write_key: read(this.cipherSuite.Cipher.keyLength),
            server_write_key: read(this.cipherSuite.Cipher.keyLength),
            client_write_IV: read(this.cipherSuite.Cipher.fixedIvLength),
            server_write_IV: read(this.cipherSuite.Cipher.fixedIvLength),
        };
    }
}
