// node-dtls-client/build/TLS/TypeSpecs.js
export function getPrimitiveSize(spec) {
    return +spec.size.substr("uint".length);
}
export var Vector;
(function (Vector) {
    function isVariableLength(spec) {
        return spec.maxLength !== spec.minLength;
    }
    Vector.isVariableLength = isVariableLength;
})(Vector || (Vector = {}));
export var Buffer;
(function (Buffer) {
    function isVariableLength(spec) {
        return spec.maxLength !== spec.minLength;
    }
    Buffer.isVariableLength = isVariableLength;
})(Buffer || (Buffer = {}));
// Shortcuts:
export const define = {
    Enum: (size, enumType) => ({ type: "enum", size, enumType }),
    Number: (size) => ({ type: "number", size }),
    Struct: (structType) => ({
        type: "struct",
        structType: structType,
    }),
    Vector: (itemSpec, minLength = 0, maxLength = minLength, optional = false) => ({
        type: "vector",
        itemSpec,
        minLength, maxLength,
        optional,
    }),
    Buffer: (minLength = Number.POSITIVE_INFINITY, maxLength = minLength) => ({
        type: "buffer",
        minLength, maxLength,
    }),
};
export const uint8 = Object.freeze(define.Number("uint8"));
export const uint16 = Object.freeze(define.Number("uint16"));
export const uint24 = Object.freeze(define.Number("uint24"));
export const uint32 = Object.freeze(define.Number("uint32"));
export const uint48 = Object.freeze(define.Number("uint48"));
export const uint64 = Object.freeze(define.Number("uint64"));
